Code/Framework/AzFramework/AzFramework/Terrain/TerrainDataRequestBus.h

```diff
@@ -26,6 +26,100 @@ namespace AzFramework
+        };
+
+        //! A JobContext used to run jobs spawned by calls to the various Query*Async functions.
+        class TerrainJobContext : public AZ::JobContext
+        {
+        public:
+            TerrainJobContext(AZ::JobManager& jobManager, int numJobsToComplete)
+                : JobContext(jobManager)
+                , m_numJobsToComplete(numJobsToComplete)
+            {
+            }
+
+            // When a terrain job context is cancelled, all associated
+            // jobs are still guaranteed to at least begin processing,
+            // and if any QueryAsyncParams::m_completionCallback was
+            // set it's guaranteed to be called even in the event of a
+            // cancellation. If a job only begins processing after its
+            // associated job context has been cancelled, no processing
+            // will occur and the callback will be invoked immediately,
+            // otherwise the job may either run to completion or cease
+            // processing early; the callback is invoked in all cases,
+            // provided one was specified with the original request.
+            void Cancel()
+            {
+                m_isCancelled = true;
+            }
+
+            // Was this TerrainJobContext cancelled?
+            bool IsCancelled() const
+            {
+                return m_isCancelled;
+            }
+
+            // Called by the TerrainSystem when a job associated with
+            // this TerrainJobContext completes. Returns true if this
+            // was the final job to be completed, or false otherwise.
+            bool OnJobCompleted()
+            {
+                return (--m_numJobsToComplete == 0);
+            }
+
+        private:
+            AZStd::atomic_int m_numJobsToComplete = 0;
+            AZStd::atomic_bool m_isCancelled = false;
+        };
+
+        //! Alias for an optional callback function to invoke when the various Query*Async functions complete.
+        //! The TerrainJobContext, returned from the original Query*Async function call, is passed as a param
+        //! to the callback function so it can be queried to see if the job was cancelled or completed normally.
+        typedef AZStd::function<void(AZStd::shared_ptr<TerrainJobContext>)> QueryAsyncCompleteCallback;
+
+        //! A parameter group struct that can optionally be passed to the various Query*Async API functions.
+        struct QueryAsyncParams
+        {
+            //! The default minimum number of positions per async terrain request job.
+            static constexpr int32_t MinPositionsPerJobDefault = 8;
+
+            //! The maximum number of jobs which async terrain requests will be split into.
+            //! This is not the value itself, rather a constant that can be used to request
+            //! the work be split into the maximum number of job manager threads available.
+            static constexpr int32_t NumJobsMax = -1;
```

Comment:
I think the problem here might be the documentation, not the definition. NumJobsMax and NumJobsDefault are values that you can set m_desiredNumberOfJobs to.  NumJobsMax isn't meant to be an absolute maximum, it's a conceptual maximum based on the number of cores & job threads each specific PC has. It's a constexpr because there's no need to modify it, it's a marker, not a valid number of jobs.

Now, there are many ways to _implement_ this concept - using the value -1, the value numeric_limits::max, a separate bool, a separate enum, etc, but they all accomplish the same goal. 

With the understanding that it's a value passed in via m_desiredNumberOfJobs, you can see that it's not actually ignored in CalculateMaxJobs(), it's specifically used on this line:
```
const int32_t numJobsMax = (numJobsDesired > 0) ? AZStd::min(numWorkerThreads, numJobsDesired) : numWorkerThreads;
```

Reply:
Updated the documentation to try and describe this better.

Reply:
What I meant by "ignored" is when set to -1, (numJobsDesired > 0) fails the test so that value isn't used. If it's meant to be a marker instead of an actual number it's being used in a kinda confusing way. Anyway, you didn't actually change any of this in your review so not actually an issue for this PR.

Reply:
new comment definitely helps 👍